#  Antibiotic resistance genes analyses in *Burkholderia pseudomallei*

This repository is for the analyses of Antibiotic Resistance Genes (ARGs) evolutionary dynamics in *B. pseudomallei* using a global collection of the bacteria 3,341 whole-genome sequences from the public databases. This project is part of the MSc Computational methods in ecology and evolution at Imperial College London.

## Description

This project aims to identify both novel and characterised ARGs, and the prevalence of resistant variants in a global population of *B. pseudomallei*. Furthermore, I hope to disentangle how the resistance genes and variants bring about, and their association to human activities such as antibiotic treatment in clinical settings. This project utilised the combination of R and shell script in the analyses. The directories are listed as; data, code, results, and sandbox. 

## Getting Started

### Dependencies
All the softwares and packages used in this project are listed. The dependencies for installation are according to the recommended guideline of each softwares.
* Softwares:
    * [Unicycler](https://github.com/rrwick/Unicycler)
    * [Quast](https://github.com/ablab/quast)
    * [Prokka](https://github.com/tseemann/prokka)
    * [PopPUNK](https://poppunk.readthedocs.io/en/latest/)
    * [Mafft](https://mafft.cbrc.jp/alignment/software/)
    * [SNP-sites](https://github.com/sanger-pathogens/snp-sites)
    * [IQ-TREE](http://www.iqtree.org)
    * [ARIBA](https://github.com/sanger-pathogens/ariba)
    * [Gubbins](https://github.com/nickjcroucher/gubbins)
    * [PAML](https://github.com/abacus-gene/paml)
    * [BactDating](https://github.com/xavierdidelot/BactDating)
* R version 4.2.0
* R packages: ggplot2, tidyverse, ape, phytools, reshape2, treeio, ggtree, chopper (fas2phy), doParallel

### Code

## Author

Chalita Chomkatekaew
[@CChomkatekaew](https://twitter.com/CChomkatekaew)

## Acknowledgments